File: code/cpp/ch22/practice/practice_03_sliding_window_max.cpp
/*
 * Practice 3: Sliding Window Maximum
 * Chapter 22: Stacks & Queues — Order Matters
 *
 * PROBLEM: Given array and window size k, return max of each window.
 * EXAMPLES: solve({1,3,-1,-3,5,3,6,7}, 3) -> {3,3,5,5,6,7}
 * INSTRUCTIONS: Replace the body of solve() with your solution.
 */
#include <deque>
#include <iostream>
#include <vector>
using namespace std;

vector<int> solve(vector<int> nums, int k) {
    // TODO: Replace this with your solution
    return {};
}

int main() {
    int n, k;
    cin >> n;
    vector<int> nums(n);
    for (int i = 0; i < n; i++) cin >> nums[i];
    cin >> k;
    vector<int> result = solve(nums, k);
    for (int x : result) cout << x << " ";
    cout << endl;
    return 0;
}

File: code/python/ch22/practice/practice_03_sliding_window_max.py
"""
Practice 3: Sliding Window Maximum
======================================
Chapter 22: Stacks & Queues — Order Matters

PROBLEM
-------
Given an array of integers and a window size k, return the maximum value
in each sliding window of size k as the window moves left to right.

CONSTRAINTS
-----------
- 1 <= len(nums) <= 10^5
- -10^4 <= nums[i] <= 10^4
- 1 <= k <= len(nums)

EXAMPLES
--------
Input: nums = [1,3,-1,-3,5,3,6,7], k = 3
Output: [3,3,5,5,6,7]

HINT
----
Use a deque storing indices in decreasing order of their values.

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
"""


def solve(nums: list[int], k: int) -> list[int]:
    """Return the maximum in each sliding window of size k."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    nums = list(map(int, input().strip().split()))
    k = int(input().strip())
    print(solve(nums, k))
